core/src/main/scala/org/apache/spark/crypto/CryptoCodec.scala

```diff
@@ -0,0 +1,96 @@
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.spark.crypto
+
+import scala.reflect.runtime.universe
+
+import org.apache.spark.{Logging, SparkConf}
+import org.apache.spark.crypto.CommonConfigurationKeys.SPARK_SECURITY_CRYPTO_CIPHER_SUITE_DEFAULT
+import org.apache.spark.crypto.CommonConfigurationKeys.SPARK_SECURITY_CRYPTO_CIPHER_SUITE_KEY
+import org.apache.spark.crypto.CommonConfigurationKeys.SPARK_SECURITY_CRYPTO_CODEC_CLASSES_AES_CTR_NOPADDING_KEY
+
+/**
+ * Crypto codec class, encapsulates encryptor/decryptor pair.
+ */
+abstract class CryptoCodec() {
+  /**
+   *
+   * @return the CipherSuite for this codec.
+   */
+  def getCipherSuite(): CipherSuite
+
+  /**
+   * This interface is only for Counter (CTR) mode. Generally the Encryptor
+   * or Decryptor calculates the IV and maintain encryption context internally.
+   * For example a {@link javax.crypto.Cipher} will maintain its encryption
+   * context internally when we do encryption/decryption using the
+   * Cipher#update interface.
+   * <p/>
+   * The IV can be calculated by combining the initial IV and the counter with
+   * a lossless operation (concatenation, addition, or XOR).
+   * @see http://en.wikipedia.org/wiki/Block_cipher_mode_of_operation#Counter_.28CTR.29
+   * @param initIV
+   * @param counter
+   * @param iv initialization vector
+   */
+  def calculateIV(initIV: Array[Byte], counter: Long, iv: Array[Byte])
+
+  /**
+   * @return Encryptor the encryptor
+   */
+  def createEncryptor: Encryptor
+
+  /**
+   * @return Decryptor the decryptor
+   */
+  def createDecryptor: Decryptor
+
+  /**
+   * Generate a number of secure, random bytes suitable for cryptographic use.
+   * This method needs to be thread-safe.
+   * @param bytes byte array to populate with random data
+   */
+  def generateSecureRandom(bytes: Array[Byte])
+}
+
+object CryptoCodec extends Logging {
+  def getInstance(conf: SparkConf): CryptoCodec = {
+    val name = conf.get(SPARK_SECURITY_CRYPTO_CIPHER_SUITE_KEY,
+      SPARK_SECURITY_CRYPTO_CIPHER_SUITE_DEFAULT)
+    getInstance(conf, CipherSuite.apply(name))
+  }
+
+  def getInstance(conf: SparkConf, cipherSuite: CipherSuite): CryptoCodec = {
+    getCodecClasses(conf, cipherSuite).toIterator.map {
+      case name if name == classOf[JceAesCtrCryptoCodec].getName => new JceAesCtrCryptoCodec(conf)
```

Comment:
It feels a little weird to have all this configuration if you're just going to enforce a single specific configuration. The same thing happens in the `checkCodec()` method later on where you enforce a specific cipher suite.

Are you planning on adding more things here later? If so, could you add comments explaining that this will change?